From 03177682fb0d1c808a01c57439f030d9392df915 Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sat, 27 Sep 2014 21:13:42 +0800
Subject: [PATCH 0125/5242] MLK-9638-4 usb: ehset: test: start timer for B
 host switch role back

commit  84023b37917660c7b2aaa158274b57dc824f7adb from
https://source.codeaurora.org/external/imx/linux-imx.git

After B device as host enumerates A peripheral and sets configuration,
B host should start to hand host role back to A device when this timer
expires.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 3f4a8c8a20ef69721f9e7886b8d245036d41d91a)
(cherry picked from commit 72b535f92ea078d8c3bd73186763bcb2f39106bb)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/misc/ehset.c    |    9 ++++++++-
 include/linux/usb/otg-fsm.h |    1 +
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/misc/ehset.c b/drivers/usb/misc/ehset.c
index ffea8015..8c6a238 100644
--- a/drivers/usb/misc/ehset.c
+++ b/drivers/usb/misc/ehset.c
@@ -111,9 +111,16 @@ static int ehset_probe(struct usb_interface *intf,
 
 		break;
 	case TEST_OTG_TEST_DEVICE_SUPPORT:
-		if (!fsm || dev->bus->is_b_host)
+		if (!fsm)
 			return ret;
 
+		/* B host enumerate test device */
+		if (dev->bus->is_b_host) {
+			otg_add_timer(fsm, B_TST_SUSP);
+			ret = 0;
+			break;
+		}
+
 		mutex_lock(&fsm->lock);
 		fsm->tst_maint = 1;
 		if (le16_to_cpu(dev->descriptor.bcdDevice) & 0x1)
diff --git a/include/linux/usb/otg-fsm.h b/include/linux/usb/otg-fsm.h
index 5524d01..fa63e5f 100644
--- a/include/linux/usb/otg-fsm.h
+++ b/include/linux/usb/otg-fsm.h
@@ -57,6 +57,7 @@ enum otg_fsm_timer {
 	A_DP_END,
 	A_TST_MAINT,
 	B_SRP_REQD,
+	B_TST_SUSP,
 
 	NUM_OTG_FSM_TIMERS,
 };
-- 
1.7.9.5

